src/sentry/sentry_metrics/indexer/cloudspanner_model.py

```diff
@@ -0,0 +1,72 @@
+from collections import namedtuple
+from enum import Enum
+from typing import Sequence
+
+_COLUMNS = [
+    "id",
+    "decoded_id",
+    "string",
+    "organization_id",
+    "date_added",
+    "last_seen",
+    "retention_days",
+]
+
+SpannerIndexerModel = namedtuple("SpannerIndexerModel", _COLUMNS)
+
+
+def get_column_names() -> Sequence[str]:
+    return _COLUMNS
+
+
+class CloudSpannerInsertMode(Enum):
+    """
+    The method to use when inserting data into CloudSpanner.
+    """
+
+    DML = 1
+    MUTATION = 2
+
+
+class CloudSpannerDBAccessor:
+    """
+    Provides methods to perform INSERT's and READ's on CloudSpanner.
+    """
+
+    def __init__(self, database, table_name: str, insert_mode: CloudSpannerInsertMode) -> None:
+        self.__database = database
+        self.__table_name = table_name
+        self.__insert = (
+            self.__insert_using_mutation
+            if insert_mode == CloudSpannerInsertMode.MUTATION
+            else self.__insert_using_dml
+        )
+
+    def __insert_using_dml(self, models: Sequence[SpannerIndexerModel]) -> None:
+        """
+        Insert data using DML. Raise any errors that occur so that the
+        application can handle them.
+        """
+
+        def insert_dml(transaction):
+            """
+            Inserts data on a database table in a transaction context.
+            """
+            transaction.insert(self.__table_name, columns=get_column_names(), values=models)
+
+        self.__database.run_in_transaction(insert_dml)
+
+    def __insert_using_mutation(self, models: Sequence[SpannerIndexerModel]) -> None:
+        """
+        Insert data using Mutation. Raise any errors that occur so that the
+        application can handle them.
+        """
+        with self.__database.batch() as batch:
+            batch.insert(
+                table=self.__table_name,
+                columns=get_column_names(),
+                values=models,
+            )
```

Comment:
I guess this will prevent us from doing a transactional read before the insert. Is that right ?

Reply:
Yes that's correct. This does a simple insert and does not meet all requirements of the indexer. Still, lets keep it for now. We can remove this later on if we don't use it.